crates/primitives/src/header.rs

```diff
@@ -261,12 +303,38 @@ impl Encodable for Header {
         // but withdrawals root is present.
         if let Some(ref base_fee) = self.base_fee_per_gas {
             U256::from(*base_fee).encode(out);
-        } else if self.withdrawals_root.is_some() {
+        } else if self.withdrawals_root.is_some() ||
+            self.blob_gas_used.is_some() ||
+            self.excess_blob_gas.is_some()
+        {
             out.put_u8(EMPTY_STRING_CODE);
         }
 
+        // Encode withdrawals root. Put empty string if withdrawals root is missing,
+        // but blob gas used is present.
         if let Some(ref root) = self.withdrawals_root {
             root.encode(out);
+        } else if self.blob_gas_used.is_some() || self.excess_blob_gas.is_some() {
+            out.put_u8(EMPTY_STRING_CODE);
+        }
+
+        // Encode blob gas used. Put empty string if blob gas used is missing,
+        // but excess blob gas is present.
+        if let Some(ref blob_gas_used) = self.blob_gas_used {
+            U256::from(*blob_gas_used).encode(out);
+        } else if self.excess_blob_gas.is_some() {
+            out.put_u8(EMPTY_STRING_CODE);
+        }
+
+        // Encode excess blob gas. If new fields are added, the above pattern will need to be
+        // repeated and placeholders added. Otherwise, it's impossible to tell _which_ fields
+        // are missing. This is mainly relevant for contrived cases where a header is created
+        // at random, for example:
+        //  * A header is created with a withdrawals root, but no base fee. Shanghai blocks are
```

Comment:
we should think of how to make this less horrible, since it's possible there will be even more header extensions in the future

Reply:
maybe different header types? like we do with transactions?

Reply:
ah, this wouldn't work because we don't have a tx type byte like we do for transactions